### ISSUE TITLE
LKMM xchg as a full barriers

### ISSUE THREAD

**Issue by db7** [MPI-SWS/genmc#22]
In LKMM, xchg should work as a full barrier according to [this paper, table 3](https://dl.acm.org/doi/pdf/10.1145/3296957.3177156):

|  Primitive           | Event |
| ----------------- | ------------------ |
| xchg_relaxed()   | R[once],W[once] |
| xchg_acquire()   | R[acquire],W[once] |
| xchg_release()   | R[once],W[release] |
| **xchg()**            | **F[mb],R[once],W[once],F[mb]** |

A use of that "feature" can be seen in Linux mcs_spinlock.h:

```c
static inline
void mcs_spin_lock(struct mcs_spinlock **lock, struct mcs_spinlock *node)
{
	struct mcs_spinlock *prev;

	/* Init node */
	node->locked =  0;
	node->next = NULL;

	/*
	 * We rely on the full barrier with global transitivity implied by the
	 * below xchg() to order the initialization stores above against any
	 * observation of @node. And to provide the ACQUIRE ordering associated
	 * with a LOCK primitive.
	 */
	prev = xchg(lock, node);
	if (likely(prev == NULL)) {
		return;
	}
	WRITE_ONCE(prev->next, node);
```

But in [lkmm.h](https://github.com/MPI-SWS/genmc/blob/1142c61028eea365c01a2d370210b6350ce2452f/include/lkmm.h#L109) `xchg()` maps to `seq_cst` implicit barriers:
```c
#define xchg(p, v)         atomic_exchange_explicit(p, v, memory_order_seq_cst)
```

If that were the case, the mcs_spinlock.h would be wrong and a release barrier would be necessary between the initialization of `node->locked` and `WRITE_ONCE(prev->next, node)` to avoid deadlocks.

I replaced `xchg()` with the following in the `lkmm.h` and it fixes this issue:
```c
#ifdef FIX_XCHG
#define xchg(p, v) ({                                                          \
	__typeof__((v)) _v_ = (v);                                             \
	(smp_mb(), _v_ = atomic_exchange_explicit(p, v, memory_order_seq_cst), \
	 smp_mb(), _v_);})
#else
#define xchg(p, v)         atomic_exchange_explicit(p, v, memory_order_seq_cst)
#endif
```

**Comment by JonasOberhauser** [comment]
`(smp_mb(), _v_ = atomic_exchange_explicit(p, v, memory_order_seq_cst), \
	 smp_mb(), _v_);})`
This should be
`(smp_mb(), _v_ = atomic_exchange_explicit(p, v, memory_order_relaxed), \
	 smp_mb(), _v_);})`
to avoid some unnecessary bookkeeping/sc acyclicity checks

**Comment by db7** [comment]
@vafeiadis this issue is actually related to the question we discussed the other day. Are all RMWs full barriers in LKMM or only `xchg`?

**Comment by db7** [comment]
> `(smp_mb(), _v_ = atomic_exchange_explicit(p, v, memory_order_seq_cst), \ smp_mb(), _v_);})`
> This should be
> `(smp_mb(), _v_ = atomic_exchange_explicit(p, v, memory_order_relaxed), \ smp_mb(), _v_);})`
> to avoid some unnecessary bookkeeping/sc acyclicity checks

True, thanks.

**Comment by michaliskok** [comment]
> In LKMM, xchg should work as a full barrier according to [this paper, table 3](https://dl.acm.org/doi/pdf/10.1145/3296957.3177156):
> 
> Primitive	Event
> xchg_relaxed()	R[once],W[once]
> xchg_acquire()	R[acquire],W[once]
> xchg_release()	R[once],W[release]
> **xchg()**	**F[mb],R[once],W[once],F[mb]**

You are right; I had copied some definitions from herd's def file, but didn't realize these definitions also generate fences. Thanks for catching that.

> @vafeiadis this issue is actually related to the question we discussed the other day. Are all RMWs full barriers in LKMM or only `xchg`?

According to the Linux-kernel implementation (and the re-interpretation of the def file), all SC value-returning atomics should act as full barriers.

**Comment by michaliskok** [comment]
Fixed in v0.6.1.
